Add FillItem method.

deepakjois opened the pull request:
Fixes #166.

We now expose a public FillItem method of KVItem struct which allows for filling values selectively during key-only iteration.

A few caveats here:

1. This requires a valid item returned during a Badger iteration. If the item is not valid, the behavior is undefined.

2. I chose acquire a lock before calling the private `fillitem` method. This may not be strictly necessary. But I wanted to avoid a misbehaving caller which may call this method even from a key-value iteration. Calling this method from an ongoing key-value iteration can interact with the prefetching.

Please let me know your thoughts, @manishrjain @srh

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/badger/175)
<!-- Reviewable:end -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/dgraph-io/badger?pullRequest=175) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/dgraph-io/badger?pullRequest=175) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/dgraph-io/badger?pullRequest=175) it.</sub>

deepakjois commented:
I think this might be cleaner if we refactor the code a little bit, to keep the waitgroup logic in a single place.

srh commented:
I don't follow -- what lock do you acquire?

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/175#-:-Ks82BXflv6I3lsDyn2a:bazs5e1)*
<!-- Sent from Reviewable.io -->

deepakjois commented:
I mean I use the waitgroup methods `wg.Wait` to wait for any pending fill to finish, and `wg.Add` before I call the fillItem method. It seems necessary to do this now that we make FillItem public, as we dont control how and when it is called.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/175#-:-Ks83yLAgh-Ma1TmSpqo:b-j8witw)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Looks good to me, but do add unit tests and verify that this behaves as expected and doesn't add any deadlocks etc. Leave it to @srh for the final approval.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/175#-:-Ks8Tz2QkXpNVChoIjR8:bv6fsop)*
<!-- Sent from Reviewable.io -->

deepakjois commented:
I will add a test.

srh commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks broke.

---

*[iterator.go, line 179 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/175#-KsBYqCo2PiB2lTOeBQT:-KsBYqCo2PiB2lTOeBQU:b-q68swf) ([raw file](https://github.com/dgraph-io/badger/blob/d27560fea8184a94fb403692a66e3fb5e9bf5a4f/iterator.go#L179)):*
> ```Go
> 	item.vptr = y.Safecopy(item.vptr, vs.Value)
> 	if it.opt.FetchValues {
> 		item.wg.Wait() // wait for any pending fill
> ```

This should be unnecessary, right?  We only call `fill` on items for which the waitgroup's counter is zero.  They're returned by `newItem`, and it's true of both items in `it.waste` and newly constructed items that `wg`'s counter must be zero.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/175)*
<!-- Sent from Reviewable.io -->

deepakjois commented:
This is now ready for another review.
* @srh, addressed your comment.
* Made another change (as a separate commit) to reset items from waste list before giving them back.

srh commented:
Review status: 1 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[iterator.go, line 122 at r2](https://reviewable.io:443/reviews/dgraph-io/badger/175#-KsGAfNST9OD4JCkDPr4:-KsGAfNTW2TwfLplBnQD:b1483ta) ([raw file](https://github.com/dgraph-io/badger/blob/7a33ad4ae6bde7158d4af219a24b4d7a86c2b1fa/iterator.go#L122)):*
> ```Go
> 		item = &KVItem{slice: new(y.Slice)}
> 	}
> 	// Clean up old values
> ```

We deliberately reuse the slices pointed at by key and vptr, in `fill`, using `Safecopy`.  So setting the fields to nil will have the effect of making us reallocate the slices and also re-initialize their content to zero.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/175)*
<!-- Sent from Reviewable.io -->

deepakjois commented:
Review status: 1 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[iterator.go, line 122 at r2](https://reviewable.io:443/reviews/dgraph-io/badger/175#-KsGAfNST9OD4JCkDPr4:-KsH2cUJo5UF_5B2YDSx:b-5dpdht) ([raw file](https://github.com/dgraph-io/badger/blob/7a33ad4ae6bde7158d4af219a24b4d7a86c2b1fa/iterator.go#L122)):*
<details><summary><i>Previously, srh (Sam Hughes) wrote…</i></summary><blockquote>

We deliberately reuse the slices pointed at by key and vptr, in `fill`, using `Safecopy`.  So setting the fields to nil will have the effect of making us reallocate the slices and also re-initialize their content to zero.
</blockquote></details>

Yes, I missed that. I should not set item.key and item.vptr to nil. But setting everything else to nil should be fine. In fact, item.val _needs_ to be set to nil.

Maybe it is better if I just set item.val explicitly to nil in the Iterator.fill method below. I think it will be a cleaner code change.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/175)*
<!-- Sent from Reviewable.io -->

deepakjois commented:
Review status: 1 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[iterator.go, line 122 at r2](https://reviewable.io:443/reviews/dgraph-io/badger/175#-KsGAfNST9OD4JCkDPr4:-KsH4hYixksZ_wwK3kq8:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/7a33ad4ae6bde7158d4af219a24b4d7a86c2b1fa/iterator.go#L122)):*
<details><summary><i>Previously, deepakjois (Deepak Jois) wrote…</i></summary><blockquote>

Yes, I missed that. I should not set item.key and item.vptr to nil. But setting everything else to nil should be fine. In fact, item.val _needs_ to be set to nil.

Maybe it is better if I just set item.val explicitly to nil in the Iterator.fill method below. I think it will be a cleaner code change.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/175)*
<!-- Sent from Reviewable.io -->

srh commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r1, 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/175#-:-KsMBRzSrGoPTxtVyYra:bnfp4nl)*
<!-- Sent from Reviewable.io -->
